app/ydoc-server-polyglot/src/insight.ts

```diff
@@ -0,0 +1,127 @@
+
+  const attachedSubdocs = new Map<Y.Doc, ReturnType<typeof attachProvider>>()
+
+  function onSubdocs(e: SubdocsEvent) {
+    e.loaded.forEach((subdoc) => {
+      attachedSubdocs.set(subdoc, attachProvider(url, subdoc.guid, subdoc))
+    })
+    e.removed.forEach((subdoc) => {
+      const subdocProvider = attachedSubdocs.get(subdoc)
+      attachedSubdocs.delete(subdoc)
+      if (subdocProvider != null) {
+        subdocProvider.dispose()
+      }
+    })
+  }
+
+  provider.on('sync', onSync)
+  provider.on('connection-close', onDrop)
+  provider.on('connection-error', onDrop)
+  doc.on('subdocs', onSubdocs)
+
+  function dispose() {
+    provider.disconnect()
+    provider.off('sync', onSync)
+    provider.off('connection-close', onDrop)
+    provider.off('connection-error', onDrop)
+    doc.off('subdocs', onSubdocs)
+    attachedSubdocs.forEach((subdocProvider) => {
+      subdocProvider.dispose()
+    })
+  }
+  return { provider, dispose }
+}
+
+const d = new Y.Doc()
+const project = new DistributedProject(d)
+//const provider = attachProvider('ws://[::1]:5976/project', 'index', d)
+
+function parseContents(contents: string, syncModule: Ast.MutableModule) {
+  const { code, idMapJson, metadataJson } = splitFileContents(contents)
+  const parsedIdMap = deserializeIdMap(idMapJson)
+
+  const { root, spans } = Ast.parseModuleWithSpans(code, syncModule)
+  syncModule.setRoot(root)
+  spanMap = spans
+
+  Ast.setExternalIds(syncModule, spans, parsedIdMap)
+  print(`[parseContents] idMap.entries=${parsedIdMap.entries()}`)
+}
+
+insight.on("source", function(ctx, frame) {
+  print(`[source] ctx.uri=${ctx.name}`)
+  if (ctx.name.includes(PROJECT_NAME)) {
+    print(`[source] MATCHED ctx=${JSON.stringify(ctx)}`)
+    parseContents(ctx.characters, syncModule)
+  }
+})
+
+function resultToString(result: any): string {
+  return '' + result;
+}
+
+/*
+  class IdMap {
+    private readonly rangeToExpr: Map<SourceRange, ExternalId>
+  }
+  interface SpanMap {
+    nodes: Map<NodeKey, Ast[]>
+    tokens: Map<TokenKey, Token>
+  }
+  type NodeKey = TokenKey = SourceRangeKey
+ */
+insight.on("return", function(ctx, frame) {
+  print(`[return] ctx='${JSON.stringify(ctx)}'`)
+  print(`[return] frame='${JSON.stringify(frame)}'`)
+  print(`[return] ctx.returnValue='${ctx.returnValue(frame)}'`)
+
+  const p = project
+  const result = ctx.returnValue(frame)
+  if (result && spanMap) {
+    const key = Ast.nodeKey(ctx.charIndex, ctx.charEndIndex - ctx.charIndex)
+    const asts = spanMap.nodes.get(key)
+    if (asts) {
+      const resultValue = resultToString(result)
+      for (const ast of asts) {
+        const editAst = syncModule.getVersion(ast)
+        if (editAst.widgetMetadata(RETURN_VALUE_KEY) !== resultValue) {
+          print(`[return] setWidgetMetadata ${ast.code()} : ${resultValue}`)
+          editAst.setWidgetMetadata(RETURN_VALUE_KEY, resultValue)
+        }
+      }
+    }
+  }
+}, {
+  expressions: true,
+  statements: false,
```

Comment:
I believe `n = 42` would be a `statement` - e.g. if you want to observe when "IDE nodes" are being assigned to, you may want to observe statements.